backend/msgpack/write_op.go

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package msgpack
+
+import (
+	"github.com/m3db/m3metrics/protocol/msgpack"
+	"github.com/m3db/m3x/pool"
+)
+
+type writeOp struct {
+	refCount
+
+	buf  msgpack.Buffer
+	pool *writeOpPool
+}
+
+func newWriteOp(pool *writeOpPool) *writeOp {
+	w := &writeOp{pool: pool}
+	w.refCount.SetDestructor(w.Close)
+	return w
+}
+
+func (w *writeOp) Reset(pending int, buf msgpack.Buffer) {
+	w.refCount.SetRefCount(pending)
+	w.buf = buf
+}
+
+func (w *writeOp) Complete(err error) {
+	w.DecRef()
+}
+
+func (w *writeOp) Close() {
+	if w.buf != nil {
+		w.buf.Close()
+	}
+	if w.pool != nil {
+		w.pool.Put(w)
+	}
+}
+
+type writeOpPool struct {
+	pool pool.ObjectPool
+}
+
+func newWriteOpPool(opts pool.ObjectPoolOptions) *writeOpPool {
+	p := pool.NewObjectPool(opts)
+	return &writeOpPool{pool: p}
+}
+
+func (p *writeOpPool) Init() {
+	p.pool.Init(func() interface{} {
+		return newWriteOp(p)
+	})
+}
+
+func (p *writeOpPool) Get() *writeOp {
+	w := p.pool.Get().(*writeOp)
+	return w
+}
+
+func (p *writeOpPool) Put(w *writeOp) {
+	w.Reset(0, nil)
```

Comment:
The `writeOpPool` is already unexported, so it's only used inside this package. I can add a comment to make that more clear, sure thing.